exonum/src/node/state.rs

```diff
@@ -92,6 +92,16 @@ pub struct State {
 
     // Cache that stores transactions before adding to persistent pool.
     tx_cache: BTreeMap<Hash, Verified<AnyTx>>,
+
+    // An in-memory set of transaction hashes, rejected by a node
+    // within block.
+    //
+    // Those transactions are stored to be known if some node will propose a block
+    // with one of them, so node could lookup for it.
+    //
+    // This set is cleared every block.
+    invalid_txs: HashSet<Hash>,
```

Comment:
This may be a vector for DoS attacks by memory exhaustion. However, if implementing a more complex solution (e.g., using an LRU cache with the bounded size) is significantly more complex, I'd go with the simpler solution for now and improve it incrementally (we have a "DoS resistance" milestone).